### PR TITLE
[onnx export] Add broadcast to matmul shape inference

### PR DESCRIPTION
**Pull request by lazycal**
When exporting a model to onnx format, the current shape inference logic of matmul does not take broadcast into consideration: 

https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx/shape_type_inference.cpp#L974-L982 

It simply uses the larger-dimensional input as the output shape (for the broadcast dimensions). A reproducible code snippet:
```python
import torch
import numpy as np


class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()

    def forward(self, x, y):
        return torch.matmul(x, y)


x_np = np.random.rand(1, 5, 8, 16).astype(np.float32)
y_np = np.random.rand(3, 1, 16, 4).astype(np.float32)
filename = 'output.onnx'

dummy_inputs = [torch.tensor(x_np), torch.tensor(y_np)]
print(Model()(*dummy_inputs).shape)
torch.onnx.export(
    Model(), tuple(dummy_inputs),
    filename,
    input_names=['x', 'y'],
    opset_version=14)

# uncomment the following lines to see what onnx shape inferencer complains about the shapes conflict
# from onnx import shape_inference, load
# inferred_model = shape_inference.infer_shapes(load("output.onnx")) 
```
The correct output shape should be [3,5,8,4], but got 1,5,8,4 from the generated model:
image

This PR fixes this issue by reusing the same code of broadcast from the function `ProcessBroadcastNode`.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 1]
```diff
@@ -800,54 +800,59 @@ void SetShapeValueFromListConstructNode(Node* lc_node) {
...
+          std::max(static_0_sz, static_1_sz));
+    } else if (!is_static_0 && !is_static_1) {
+      if (ss_shape_0.value() == ss_shape_1.value()) {
+        final_shape[rank_max - 1 - idx] = ss_shape_0;
+      }
+    }
+  }
+
+  if (rank_0 < rank_1) {
+    for (auto idx = rank_min; idx < rank_max; idx++) {
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:834]:
```suggestion
    for (size_t idx = rank_min; idx < rank_max; idx++) {
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 2]
```diff
@@ -800,54 +800,59 @@ void SetShapeValueFromListConstructNode(Node* lc_node) {
...
+      auto shape_idx = rank_max - 1 - idx;
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:835]:
```suggestion
      size_t shape_idx = rank_max - 1 - idx;
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 3]
```diff
@@ -800,54 +800,59 @@ void SetShapeValueFromListConstructNode(Node* lc_node) {
...
+      final_shape[shape_idx] = input_shape_value_1[shape_idx];
+    }
+  } else {
+    for (auto idx = rank_min; idx < rank_max; idx++) {
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:839]:
```suggestion
    for (size_t idx = rank_min; idx < rank_max; idx++) {
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 4]
```diff
@@ -800,54 +800,59 @@ void SetShapeValueFromListConstructNode(Node* lc_node) {
...
+      auto shape_idx = rank_max - 1 - idx;
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:840]:
```suggestion
      size_t shape_idx = rank_max - 1 - idx;
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 5]
```diff
@@ -968,25 +973,15 @@ void ProcessMatMulNode(Node* n) {
...
-      for (auto idx = 0; idx < rank_1 - 2; idx++) {
-        final_shape.emplace_back(input_shape_value_1[idx]);
-      }
-    }
-    final_shape.emplace_back(input_shape_value_0[rank_0 - 2]);
-    final_shape.emplace_back(input_shape_value_1[rank_1 - 1]);
-    if (is_rank_0_1) {
-      final_shape.erase(final_shape.begin());
+    auto final_shape = Broadcast(
+      std::vector<::c10::ShapeSymbol>(input_shape_value_0.begin(), input_shape_value_0.end() - 2),
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:977]: Please add a comment explaining the `- 2`.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 6]
```diff
@@ -968,25 +973,15 @@ void ProcessMatMulNode(Node* n) {
...
+      std::vector<::c10::ShapeSymbol>(input_shape_value_1.begin(), input_shape_value_1.end() - 2)
+    );
+    if (!is_rank_0_1) {
+      final_shape.emplace_back(input_shape_value_0[rank_0 - 2]);
     }
-    if (is_rank_1_1) {
-      final_shape.pop_back();
+    if (!is_rank_1_1) {
+      final_shape.emplace_back(input_shape_value_1[rank_1 - 1]);
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:984]: Please add a comment explaining the `- 1`.

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Please add a test that fails before this change and passes afterwards.

Perhaps in test/jit/test_onnx_export.py

**Comment by lazycal** [comment]
@garymm Thanks for reviewing this PR! I have made the changes according to your suggesetion. One difference is that I put the unit test in test/onnx/test_pytorch_onnx_shape_inference.py. Can you take a look again?

File: test/onnx/test_pytorch_onnx_shape_inference.py [hunk 7]
```diff
@@ -106,5 +106,33 @@ def test_slice(self):
         slice = g.op("Slice", input, start_input, end, axis, step)
         self.run_test(g, slice.node(), expect_tensor(None, shape=(None, None)))
 
+    def test_bcast_matmul(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_shape_inference.py:109]: please spell out "broadcast"

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 8]
```diff
@@ -955,6 +960,11 @@ void ProcessMatMulNode(Node* n) {
     auto input_shape_value_1 = input_shape_1.sizes().value();
     size_t rank_0 = input_shape_value_0.size();
     size_t rank_1 = input_shape_value_1.size();
+    // the matmul supports the rank of either or both inputs to be 1
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:963]:
```suggestion
    // Inputs of rank 1 handled as in numpy.matmul:
    // https://numpy.org/doc/stable/reference/generated/numpy.matmul.html
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 9]
```diff
@@ -955,6 +960,11 @@ void ProcessMatMulNode(Node* n) {
...
+    // the behavior is the same as appending a 1 to the input(s), do the matmul and remove the appended 1(s)
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:964]: Please delete the rest of the comment.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 8 above)
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:963]:
```suggestion
    // Handle inputs of rank 1 just like numpy.matmul:
    // https://numpy.org/doc/stable/reference/generated/numpy.matmul.html
```

File: test/onnx/test_pytorch_onnx_shape_inference.py (hunk 7 above)
Comment (garymm) [inline test/onnx/test_pytorch_onnx_shape_inference.py:109]:
```suggestion
    def test_broadcast_matmul(self):
```

**Review by garymm — APPROVED** [review]
Thanks!

**Comment by lazycal** [comment]
Thanks. The comment and name are changed as you suggested.

**Comment by garymm** [comment]
@msaroufim could you please import this? (the new merge bot doesn't work for this case it seems)

**Comment by garymm** [comment]
@lazycal if you open a new PR with the same content against the branch `onnx_ms_1`, then I can merge it for you, and it will eventually make it into `master`.

**Comment by lazycal** [comment]
@garymm I changed the PR instead to merge into `onnx_ms_1`. If this is not OK I can also make a new PR.

**Comment by garymm** [comment]
@lazycal please resolve the merge conflicts and then I will merge this.

**Pull request merged**